### ISSUE TITLE
Undo should undo autoformatting

### ISSUE THREAD

**Issue by jkpubsrc** [mono/monodevelop#4090]
A very convenient feature of Microsoft Visual Studio is that an Undo operation can undo auto-formatting. This can be achieved by separating a key event from an auto formatting task: For the undo-redo-list two entries are generated instead of on. In that case if you undo something the first undo step will remove the last auto-formatting activity, the next undo will then undo the key typed.

This is something very reasonable as auto-formatting not always works perfectly. Especially if your code is not yet 100% conform to C# standards as your currently in the middle of editing it, sometimes auto-formatting will do some weired stuff. This is normal and can hardly be avoided if the code does not yet compile in the middle of an editing process. Undo is an easy way of resolving this problem: If auto-formatting get's in your way, in VS you simple can undo this. In MonoDevelop you can't as auto-formatting is not a task on it's own in the undo list.

You could consider this as feature, I'd even go that far to consider that as a bug.

**Comment by slluis** [comment]
This is the current behavior in MonoDevelop. Do you have a specific use case in which that doesn't work?

**Comment by jkpubsrc** [comment]
Ehm .... Sorry, I don't understand the question. This is because I trip over this situation quite frequently if I edit source code files. That's why I'm a bit surprised you ask for a "use case".

Whenever you write software in the IDE you encounter this behavior. For example if you have not yet figured out which settings of the auto-formatting logic are perfect for your needs. You edit your code and auto-formatting does not exactly do what you want to, then you want to undo these formatting-changes (and not YOUR editing changes). In that situation you immediately encounter the described situation.

Another use case would be that you deliberately want to deviate from the auto-formatting provided. An example for that I had yesterday, as I provided a specific hard coded configuration for my application. Basically this can be considered as a domain specific language. For easy understanding by someone else who would look at my code I provided that configuration in a table-like form. That means: I added tabs to align specific information within a single source code line in such a way that specific aspects are aligned exactly below each other. Which was difficult (!) to enter in MonoDevelop as auto-formatting can't be undone, only auto-formatting TOGETHER with your edits can be undone. But I want to keep my edits in these situations as they correct errors and improve the code during the process of software development.

(Situations like these are more frequent than you might think.)

**Comment by jkpubsrc** [comment]
By the way: If you paste preformatted text in to MonoDevelop auto-format destroys that formatting. The recommended user experience would now be to perform a single undo step to the be able to continue and do formatting yourself. This is not possible ein MonoDevelop. If you undo the whole insert-and-formatting-step is undone. So your only way currently is to accept auto-formatting to modify the entire clip of text/data you inserted and then do ALL formatting again by hand.

If you separate the insert operation and the auto formatting operations into two different steps within the edit component (as it is in other IDEs/editors to my experience) this inconvenient behavior is gone and user experience is improved.

**Comment by jkpubsrc** [comment]
Damn. I'm closing this issue. So much effort for nothing: **There is an option for that in the preferences.**

Hell I'm wondering why this option is not the default! I'm really wondering what sense it should make, allowing the user a choice here as I don't see ANY reason why it edit-and-formatting should be a single undo step instead of two! It doesn't make any sense!

**Comment by slluis** [comment]
I agree this option should be enabled by default. I thought it already was. /cc @mkrueger

**Comment by mkrueger** [comment]
I checked the source the option is on by default however that wasn't always be the case. So it is possible that the setting was wrongly taken from an older version.

However my pov is that we should remove the option. Generating an undo step for all IDE automatically made changes is an industry standard accross all IDEs. That is the case for ~20 years now AFAIR.

That is one of these options where is simply no need for - my2cents.
